utils/wasm-gen/src/lib.rs

```diff
@@ -788,63 +855,157 @@ impl<'a> WasmGen<'a> {
-                &data,
-                memory_pages,
-                source_call_index,
-                &mut offsets,
-            );
-            for _ in 0..data.sys_call_amount {
-                module = self.insert_instructions_in_random_place(module, &instructions);
-            }
-        }
+        let func_type = send_from_reservation_call_info.func_type();
+
+        let func_signature = builder::signature()
+            .with_params(func_type.params().iter().copied())
+            .with_results(func_type.results().iter().copied())
+            .build_sig();
+
+        let memory_size_in_bytes = memory_pages.memory_size();
+        let reserve_gas_result_ptr = memory_size_in_bytes.saturating_sub(MEMORY_VALUE_SIZE) as i32; // Pointer to the LengthWithHash struct
+        let rid_pid_value_ptr = reserve_gas_result_ptr + size_of::<Length>() as i32;
+        let pid_value_ptr = reserve_gas_result_ptr + size_of::<LengthWithHash>() as i32;
+        let reservation_send_result_ptr = pid_value_ptr + size_of::<HashWithValue>() as i32;
+
+        let func_instructions = Instructions::new(vec![
+            Instruction::GetLocal(4),                      // Amount of gas to reserve
+            Instruction::GetLocal(5),                      // Duration of the reservation
+            Instruction::I32Const(reserve_gas_result_ptr), // Pointer to the LengthWithHash struct
+            Instruction::Call(reserve_gas_call_data.call_index),
+            Instruction::I32Const(reserve_gas_result_ptr), // Pointer to the LengthWithHash struct
+            Instruction::I32Load(2, 0),                    // Load LengthWithHash.length
+            Instruction::I32Eqz,                           // Check if LengthWithHash.length == 0
+            Instruction::If(BlockType::NoResult),          // If LengthWithHash.length == 0
+            Instruction::I32Const(pid_value_ptr), // Copy the HashWithValue struct (48 bytes) containing the recipient and value after the obtained reservation ID
+            Instruction::GetLocal(0),
+            Instruction::I64Load(3, 0),
+            Instruction::I64Store(3, 0),
+            Instruction::I32Const(pid_value_ptr),
+            Instruction::GetLocal(0),
+            Instruction::I64Load(3, 8),
+            Instruction::I64Store(3, 8),
+            Instruction::I32Const(pid_value_ptr),
+            Instruction::GetLocal(0),
+            Instruction::I64Load(3, 16),
+            Instruction::I64Store(3, 16),
+            Instruction::I32Const(pid_value_ptr),
+            Instruction::GetLocal(0),
+            Instruction::I64Load(3, 24),
+            Instruction::I64Store(3, 24),
+            Instruction::I32Const(pid_value_ptr),
+            Instruction::GetLocal(0),
+            Instruction::I64Load(3, 32),
+            Instruction::I64Store(3, 32),
+            Instruction::I32Const(pid_value_ptr),
+            Instruction::GetLocal(0),
+            Instruction::I64Load(3, 40),
+            Instruction::I64Store(3, 40),
```

Comment:
This memory copying does not look nice. I can think of making some sort of macro for this. Other thoughts?

Reply:
you have to resolve import indices/etc in that case